geopandas/geodataframe.py

```diff
@@ -1384,18 +1396,34 @@ def apply(self, func, axis=0, raw=False, result_type=None, args=(), **kwargs):
         result = super().apply(
             func, axis=axis, raw=raw, result_type=result_type, args=args, **kwargs
         )
-        if (
-            isinstance(result, GeoDataFrame)
-            and self._geometry_column_name in result.columns
-            and any(isinstance(t, GeometryDtype) for t in result.dtypes)
-        ):
-            if self.crs is not None and result.crs is None:
-                result.set_crs(self.crs, inplace=True)
+
+        if self._geometry_column_name in result.columns:
+            # Reconstruct gdf if it was lost by apply
+            if isinstance(result, GeoDataFrame) is False:
+                # axis=1 apply will split GeometryDType to object, try and cast back
+                try:
+                    _ensure_geometry(result[self._geometry_column_name], crs=self.crs)
+                except TypeError:
+                    pass
+                else:
+                    result = GeoDataFrame(
+                        result, geometry=self._geometry_column_name, crs=self.crs
+                    )
+            else:
+                # this type check seems broad, already raised in GH1849
+                if any(isinstance(t, GeometryDtype) for t in result.dtypes):
+                    result = result.set_geometry(self._geometry_column_name)
```

Comment:
This line is new in this branch compared to the original code. Was this also needed for one of the tests?

Reply:
Maybe I'm missing the point you were trying to make, but the else block (+1412:) is the the same as the old apply (-1387) - the if block being if apply returns a DataFrame and else being if it's a GeoDataFrame we do the same as before. But I've done this really clumsily, On review, `any(isinstance(t, GeometryDtype) for t in result.dtypes)` - which I think should be `isinstance(result[self._geometry_column_name].dtype, GeometryDtype)` (from #1955 ) is a subset of the `_ensure_geometry` check, so these cases could be brought together. Doing it separately you have a way of checking this without an exception, but it's probably not worth the difference. For context that would end up being - which is probably what I'll push
```python
        # Reconstruct gdf if it was lost by apply
        if self._geometry_column_name in result.columns:
            # axis=1 apply will split GeometryDType to object, try and cast back
            try:
                _ensure_geometry(result[self._geometry_column_name], crs=self.crs)
            except TypeError:
                pass
            else:
                result = result.set_geometry(self._geometry_column_name)
                if self.crs is not None and result.crs is None:
                    result.set_crs(self.crs, inplace=True)
```

Reply:
I specifically meant the `set_geometry` call, which I don't see in the original code? (-1392-1393) 
In this branch, `result` is already a GeoDataFrame, so wondering if it's needed to still do `set_geometry`.

Reply:
Ah, my bad, I should have seen that. Checking out 1afc2580 - which is the point where you made the original comment (I don't know how to easily refer to that in the PR now that I've pushed more commits), it is actually necessary - for one of the new tests I added `test_op_output_types::test_apply` when the geometry column is not called geometry. Without this  the geometry column name doesn't get updated for
`df[[set_geom_col_name, "geometry2"]].apply(lambda x: x)`.

Apparently that's an old edge case and not a new one:
```python
In [1]: import geopandas

In [2]: df = geopandas.read_file(geopandas.datasets.get_path("naturalearth_cities"))
In [4]: df2 = df.rename_geometry('points')
In [6]: df3 = df2[['name', 'points']].apply(lambda x:x)
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-5132eda620df> in <module>
----> 1 df3 = df2[['name', 'points']].apply(lambda x:x)
AttributeError: No geometry data set yet (expected in column 'geometry'.)
# for completeness, default geometry col is fine
In [7]: df[['name', 'geometry']].apply(lambda x:x)
Out[7]:
             name                     geometry
0    Vatican City    POINT (12.45339 41.90328)
1      San Marino    POINT (12.44177 43.93610)
2           Vaduz     POINT (9.51667 47.13372)
```
<details>

In [8]: geopandas.show_versions()

SYSTEM INFO
-----------
python     : 3.7.10 | packaged by conda-forge | (default, Feb 19 2021, 15:37:01) [MSC v.1916 64 bit (AMD64)]
executable : C:\MyPrograms\Miniconda3\envs\geometry\python.exe
machine    : Windows-10-10.0.19041-SP0

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : None
GEOS lib   : None
GDAL       : 3.2.2
GDAL data dir: None
PROJ       : 8.0.0
PROJ data dir: C:\MyPrograms\Miniconda3\envs\geometry\Library\share\proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.9.0
pandas     : 1.2.4
fiona      : 1.8.19
numpy      : 1.20.2
shapely    : 1.7.1
rtree      : 0.9.7
pyproj     : 3.0.1
matplotlib : None
mapclassify: None
geopy      : None
psycopg2   : 2.8.4 (dt dec pq3 ext lo64)
geoalchemy2: None
pyarrow    : None
pygeos     : None

</details>